### PR TITLE
add screenshotQuality

### PR DESCRIPTION
**Pull request by KazuCocoa**
assist https://github.com/appium/WebDriverAgent/pull/127

The screenshot quality affects the entire screenshot strategy so I'd add it in caps

### CODE REVIEW

File: lib/driver.js [hunk 1]
```diff
@@ -218,6 +219,7 @@ class XCUITestDriver extends BaseDriver {
         shouldUseCompactResponses: DEFAULT_SETTINGS.shouldUseCompactResponses,
         mjpegServerScreenshotQuality: DEFAULT_SETTINGS.mjpegServerScreenshotQuality,
         mjpegServerFramerate: DEFAULT_SETTINGS.mjpegServerFramerate,
+        screenshotQuality: DEFAULT_SETTINGS.screenshotQuality,
```
Comment (mykola-mokhnach) [inline lib/driver.js]: I think we should not repeat the same defaults in WDA and inside the driver. Lets only pass the setting, which value was atcually changed.

Comment (mykola-mokhnach) [reply lib/driver.js]: that was reasonable before we had our own wda fork

Comment (mykola-mokhnach) [reply lib/driver.js]: @imurchie @jlipps What do you think?

File: lib/driver.js [hunk 2]
```diff
@@ -34,7 +34,8 @@ const DEFAULT_SETTINGS = {
   shouldUseCompactResponses: true,
   elementResponseAttributes: "type,label",
   mjpegServerScreenshotQuality: 25, // https://github.com/appium/WebDriverAgent/blob/master/WebDriverAgentLib/Utilities/FBConfiguration.m#L29
-  mjpegServerFramerate: 10 // https://github.com/appium/WebDriverAgent/blob/master/WebDriverAgentLib/Utilities/FBConfiguration.m#L30
+  mjpegServerFramerate: 10, // https://github.com/appium/WebDriverAgent/blob/master/WebDriverAgentLib/Utilities/FBConfiguration.m#L30
+  screenshotQuality: 1 // https://github.com/appium/WebDriverAgent/blob/master/WebDriverAgentLib/Utilities/FBConfiguration.m#L31
```
Comment (mykola-mokhnach) [inline lib/driver.js]: these are probably not needed as well

File: README.md [hunk 3]
```diff
@@ -160,6 +160,7 @@ Differences noted here
...
 |`reduceMotion`| It allows to turn on/off reduce motion accessibility preference. Setting reduceMotion `on` helps to reduce flakiness during tests. Only fon simulators | e.g `true` |
 |`permissions`| Allows to set permissions for the specified application bundle on Simulator only. The capability value is expected to be a valid JSON string with `{"<bundleId1>": {"<serviceName1>": "<serviceStatus1>", ...}, ...}` format. It is required that `applesimutils` package is installed and available in PATH. The list of available service names and statuses can be found at https://github.com/wix/AppleSimulatorUtils. | e. g. `{"com.apple.mobilecal": {"calendar": "YES"}}` |
+|`screenshotQuality`| Changes the quality of phone display screenshots following [xctest/xctimagequality](https://developer.apple.com/documentation/xctest/xctimagequality?language=objc). Default value is `1`. `0` is the highest and `2` is the lowest quality. You can also change it via [settings](https://github.com/appium/appium/blob/master/docs/en/advanced-concepts/settings.md) command | e.g. `0`, `1`, `2` |
```
Comment (mykola-mokhnach) [inline README.md]: It worth mentioning that setting the value to `0` might cause OutOfMemory crashes on devices with hi-res screens, like iPad Pro or iPhone Plus

File: lib/driver.js (hunk 2 above)
Comment (KazuCocoa) [reply lib/driver.js]: Currently, the key/value checks available settings using here. We can see below error if we have no value here.

So, I would like to create another PR to change the method.

```
[XCUITest] Setting the quality of phone screenshot: '2'
[BaseDriver] Didn't know about setting 'screenshotQuality'. Are you sure you spelled it correctly? Proceeding anyway. Valid settings: imageMatchThreshold,fixImageFindScreenshotDims,fixImageTemplateSize,checkForImageElementStaleness,autoUpdateImageElementPosition,imageElementTapStrategy,nativeWebTap,useJSONSource,shouldUseCompactResponses,elementResponseAttributes,mjpegServerScreenshotQuality,mjpegServerFramerate
```

Comment (KazuCocoa) [reply lib/driver.js]: :note: https://github.com/appium/appium-base-driver/blob/f510db0257110a755c9f2976bf7580abc9402bc9/lib/basedriver/device-settings.js#L44

Comment (mykola-mokhnach) [reply lib/driver.js]: Then it would be better to remove line numbers from comments, since these might change in the future

**Comment by mykola-mokhnach** [comment]
@KazuCocoa Do you want to bump WDA component as a part of ths PR?

**Comment by KazuCocoa** [comment]
ah, good idea. Will do 👍

**Comment by KazuCocoa** [comment]
```
dbug MJSONWP (5007a9de) Responding to client with driver.getScreenshot() result: "/9j/4AAQSkZJRgABAQAAkACQAAD/4QCARXhpZgAATU0AKgAAAAgABQESAAMAAAAB\r\nAAEAAAEaAAUAAAABAAAASgEbAAUAAAABAAAAUgEoAAMAAAABAAIAAIdpAAQAAAAB\r\nAAAAWgAAAAAAAACQAAAAAQAAAJAAAAABAAKgAgAEAAAAAQAAAu6gAwAEAAAAAQAA\r\nBTYAAAAA/+0AOFBob3Rvc2hvcCAzLjAAOEJJTQQEAAAAAAAAOEJJTQQlAAAAAAAQ\r\n1B2M2Y8AsgTpgAmY7PhCfv/AABEIBTYC7gMBIgACEQEDEQH/xAAfAAABBQEBAQEB\r\nAQAAAAAAAAAAAQIDBAUGBwgJCgv/xAC1EAACAQMDAgQDBQUEBAAAAX0BAgMABBEF\r\nEiExQQYTUWEHInEUMoGRoQgjQrHBFVLR8CQzYnKCCQoWFxgZGiUmJygpKjQ1Njc4\r\nOTpDREVGR0hJSlNUVVZXWFlaY2RlZmdoaWpzdHV2d3h5eoOEhYaHiImKkpOUlZaX\r\nmJmaoqOkpaanqKmqsrO0tba3uLm6wsPExcbHyMnK0tPU1dbX2Nna4eLj5OXm5+jp\r\n6vHy8/T19vf4+fr/xAAfAQADAQEBAQEBAQEBAAAAAAAAAQIDBAUGBwgJCgv/xAC1\r\nEQACAQIEBAMEBwUEBAABAncAAQIDEQQFITEGEkFRB2FxEyIygQgUQpGhscEJIzNS\r\n8BVictEKFiQ04SXxFxgZGiYnKCkqNTY3ODk6Q0RFRkdISUpTVFVWV1hZWmNkZWZn\r\naGlqc3R1dnd4eXqCg4SFhoeIiYqSk5SVlpeYmZqio6Slpqeoqaqys7S1tre4ubrC\r\nw8TFxsfIycrS09TV1tfY2dri4+Tl5ufo6ery8/T19vf4+fr/2wBDAAICAgICAgQC\r\nAgQGBAQEBggGBgYGCAoICAgICAoMCgoKCgoKDAwMDAwMDAwODg4ODg4QEBAQEBIS\r\n...
info HTTP <-- GET /wd/hub/session/5007a9de-c287-4913-ac08-813ab0d5e147/screenshot 200 44 ms - 200058
info HTTP 
info HTTP --> POST /wd/hub/session/5007a9de-c287-4913-ac08-813ab0d5e147/execute
info HTTP {"script":"mobile: viewportScreenshot","args":[]}
dbug MJSONWP (5007a9de) Calling XCUITestDriver.execute() with args: ["mobile: viewportScreenshot",[],"5007a9de-c287-4913-ac08-813ab0d5e147"]
dbug XCUITest Executing command 'execute'
dbug XCUITest Taking screenshot with WDA
dbug JSONWP Proxy Matched '/screenshot' to command name 'getScreenshot'
dbug JSONWP Proxy Proxying [GET /screenshot] to [GET http://localhost:8100/session/6DED4E86-7A0B-459F-AEF2-31B03FA9DAC3/screenshot] with no body
dbug JSONWP Proxy Got response with status 200: "{\n  \"value\" : \"\\/9j\\/4AAQSkZJRgABAQAAkACQAAD\\/4QCARXhpZgAATU0AKgAAAAgABQESAAMAAAAB\\r\\nAAEAAAEaAAUAAAABAAAASgEbAAUAAAABAAAAUgEoAAMAAAABAAIAAIdpAAQAAAAB\\r\\nAAAAWgAAAAAAAACQAAAAAQAAAJAAAAABAAKgAgAEAAAAAQAAAu6gAwAEAAAAAQAA\\r\\nBTYAAAAA\\/+0AOFBob3Rvc2hvcCAzLjAAOEJJTQQEAAAAAAAAOEJJTQQlAAAAAAAQ\\r\\n1B2M2Y8AsgTpgAmY7PhCfv\\/AABEIBTYC7gMBIgACEQEDEQH\\/xAAfAAABBQEBAQEB\\r\\nAQAAAAAAAAAAAQIDBAUGBwgJCgv\\/xAC1EAACAQMDAgQDBQUEBAAAAX0BAgMABBEF\\r\\nEiExQQYTUWEHInEUMoGRoQgjQrHBFVLR8CQzYnKCCQoWFxgZGiUmJygpKjQ1Njc4\\r\\nOTpDREVGR0hJSlNUVVZXWFlaY2RlZmdoaWpzdHV2d3h5eoOEhYaHiImKkpOUlZaX\\r\\nmJmaoqOkpaanqKmqsrO0tba3uLm6wsPExcbHyMnK0tPU1dbX2Nna4eLj5OXm5+jp\\r\\n6vHy8\\/T19vf4+fr\\/xAAfAQADAQEBAQEBAQEBAAAAAAAAAQIDBAUGBwgJCgv\\/xAC1\\r\\nEQACAQIEBAMEBwUEBAABAncAAQIDEQQFITEGEkFRB2FxEyIygQgUQpGhscEJIzNS\\r\\n8BVictEKFiQ04SXxFxgZGiYnKCkqNTY3ODk6Q0RFRkdISUpTVFVWV1hZWmNkZWZn\\r\\naGlqc3R1dnd4eXqCg4SFhoeIiYqSk5SVlpeYmZqio6Slpqeoqaqys7S1tre4ubrC\\r\\nw8TFxsfIycrS09TV1tfY2dri4+Tl5ufo6ery8\\/T19vf4+fr\\/2wBDAAICAgICAgQC\\r...
dbug MJSONWP (5007a9de) Encountered internal error running command: Error: Invalid file signature
dbug MJSONWP (5007a9de)     at module.exports.Parser._parseSignature (/Users/travis/build/appium/appium-xcuitest-driver/node_modules/pngjs/lib/parser.js:50:18)
dbug MJSONWP (5007a9de)     at module.exports.ChunkStream._processRead (/Users/travis/build/appium/appium-xcuitest-driver/node_modules/pngjs/lib/chunkstream.js:174:13)
dbug MJSONWP (5007a9de)     at module.exports.ChunkStream._process (/Users/travis/build/appium/appium-xcuitest-driver/node_modules/pngjs/lib/chunkstream.js:193:14)
dbug MJSONWP (5007a9de)     at module.exports.ChunkStream.write (/Users/travis/build/appium/appium-xcuitest-driver/node_modules/pngjs/lib/chunkstream.js:61:8)
dbug MJSONWP (5007a9de)     at module.exports.ChunkStream.end (/Users/travis/build/appium/appium-xcuitest-driver/node_modules/pngjs/lib/chunkstream.js:74:10)
dbug MJSONWP (5007a9de)     at exports.PNG.PNG.end (/Users/travis/build/appium/appium-xcuitest-driver/node_modules/pngjs/lib/png.js:98:16)
dbug MJSONWP (5007a9de)     at exports.PNG.PNG.parse (/Users/travis/build/appium/appium-xcuitest-driver/node_modules/pngjs/lib/png.js:88:8)
dbug MJSONWP (5007a9de)     at parse (/Users/travis/build/appium/appium-xcuitest-driver/node_modules/appium-support/lib/image-util.js:476:11)
dbug MJSONWP (5007a9de)     at Promise._execute (/Users/travis/build/appium/appium-xcuitest-driver/node_modules/bluebird/js/release/debuggability.js:313:9)
dbug MJSONWP (5007a9de)     at Promise._resolveFromExecutor (/Users/travis/build/appium/appium-xcuitest-driver/node_modules/bluebird/js/release/promise.js:483:18)
dbug MJSONWP (5007a9de)     at new Promise (/Users/travis/build/appium/appium-xcuitest-driver/node_modules/bluebird/js/release/promise.js:79:10)
dbug MJSONWP (5007a9de)     at /Users/travis/build/appium/appium-xcuitest-driver/node_modules/appium-support/lib/image-util.js:474:16
dbug MJSONWP (5007a9de)     at Generator.next (<anonymous>)
dbug MJSONWP (5007a9de)     at asyncGeneratorStep (/Users/travis/build/appium/appium-xcuitest-driver/node_modules/@babel/runtime/helpers/asyncToGenerator.js:3:24)
dbug MJSONWP (5007a9de)     at _next (/Users/travis/build/appium/appium-xcuitest-driver/node_modules/@babel/runtime/helpers/asyncToGenerator.js:25:9)
dbug MJSONWP (5007a9de)     at /Users/travis/build/appium/appium-xcuitest-driver/node_modules/@babel/runtime/helpers/asyncToGenerator.js:32:7
dbug MJSONWP (5007a9de)     at new Promise (<anonymous>)
dbug MJSONWP (5007a9de)     at /Users/travis/build/appium/appium-xcuitest-driver/node_modules/@babel/runtime/helpers/asyncToGenerator.js:21:12
dbug MJSONWP (5007a9de)     at _base64ToImage (/Users/travis/build/appium/appium-xcuitest-driver/node_modules/appium-support/build/lib/image-util.js:407:25)
dbug MJSONWP (5007a9de)     at base64ToImage (/Users/travis/build/appium/appium-xcuitest-driver/node_modules/appium-support/build/lib/image-util.js:387:25)
dbug MJSONWP (5007a9de)     at Object.base64ToImage (/Users/travis/build/appium/appium-xcuitest-driver/node_modules/appium-support/lib/image-util.js:461:23)
dbug MJSONWP (5007a9de)     at Generator.next (<anonymous>)
dbug MJSONWP (5007a9de)     at asyncGeneratorStep (/Users/travis/build/appium/appium-xcuitest-driver/node_modules/@babel/runtime/helpers/asyncToGenerator.js:3:24)
dbug MJSONWP (5007a9de)     at _next (/Users/travis/build/appium/appium-xcuitest-driver/node_modules/@babel/runtime/helpers/asyncToGenerator.js:25:9)
dbug MJSONWP (5007a9de)     at /Users/travis/build/appium/appium-xcuitest-driver/node_modules/@babel/runtime/helpers/asyncToGenerator.js:32:7
dbug MJSONWP (5007a9de)     at new Promise (<anonymous>)
dbug MJSONWP (5007a9de)     at Object.<anonymous> (/Users/travis/build/appium/appium-xcuitest-driver/node_modules/@babel/runtime/helpers/asyncToGenerator.js:21:12)
dbug MJSONWP (5007a9de)     at Object._cropBase64Image (/Users/travis/build/appium/appium-xcuitest-driver/node_modules/appium-support/build/lib/image-util.js:383:27)
dbug MJSONWP (5007a9de)     at Object.cropBase64Image (/Users/travis/build/appium/appium-xcuitest-driver/node_modules/appium-support/build/lib/image-util.js:374:27)
dbug MJSONWP (5007a9de)     at XCUITestDriver.cropBase64Image [as getViewportScreenshot] (/Users/travis/build/appium/appium-xcuitest-driver/lib/commands/screenshots.js:160:39)
dbug MJSONWP (5007a9de)     at <anonymous>
info HTTP <-- POST /wd/hub/session/5007a9de-c287-4913-ac08-813ab0d5e147/execute 500 88 ms - 193
info HTTP 
      1) should get a cropped screenshot of the viewport without statusbar
    logging -
      types -
info HTTP --> GET /wd/hub/session/5007a9de-c287-4913-ac08-813ab0d5e147/log/types
info HTTP {}
```

**Comment by KazuCocoa** [comment]
`viewportScreenshot` expected png header, but the wda returns jpeg so it happened

If `screenshotQuality: 0`, it return png, `iVBOR` base64 encoded. `screenshotQuality: 1` or `2` returns jpeg, `/9j/4`.

**Comment by mykola-mokhnach** [comment]
Thanks, I will fix that

**Comment by KazuCocoa** [comment]
thanks! I tried https://github.com/appium/WebDriverAgent/pull/127#discussion_r235711922 a bit.

I think the later's `should get the screenshot via an mjpeg server if requested` is fine to update test cases

**Comment by mykola-mokhnach** [comment]
It looks like Travis is not very happy: https://travis-ci.org/appium/appium-xcuitest-driver/jobs/458448399

File: test/functional/basic/screenshot-e2e-specs.js [hunk 4]
```diff
@@ -29,9 +29,9 @@ describe('XCUITestDriver - screenshots - mjpeg server', function () {
     mjpegServer.close();
   });
 
-  it('should get the jpeg formatted screenshot via an mjpeg server if requested', async function () {
+  it('should get the screenshot via an mjpeg server if requested', async function () {
     const img = await driver.takeScreenshot();
-    img.indexOf('/9j/4A').should.eql(0);
+    img.indexOf('iVBOR').should.eql(0);
```
Comment (mykola-mokhnach) [inline test/functional/basic/screenshot-e2e-specs.js]: https://github.com/appium/appium-android-driver/blob/3643d870706f32604205f1418ba3497c133fec94/test/functional/commands/actions-e2e-specs.js#L66

**Pull request merged**